src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -1602,9 +1612,19 @@ class CleanupDumpTimeLambdaProxyClassTable: StackObj {
  public:
   bool do_entry(LambdaProxyClassKey& key, DumpTimeLambdaProxyClassInfo& info) {
     assert_lock_strong(DumpTimeTable_lock);
+    InstanceKlass* caller_ik = key.caller_ik();
+    if (SystemDictionaryShared::check_for_exclusion_impl(caller_ik, true)) {
+      // If the caller class is excluded, unregister all the associated lambda proxy classes so that
+      // they won't be included in the CDS archive.
+      for (int i = 0; i < info._proxy_klasses->length(); i++) {
+        SystemDictionaryShared::reset_registered_lambda_proxy_class(info._proxy_klasses->at(i));
+        info._proxy_klasses->remove_at(i);
+      }
+    }
     for (int i = 0; i < info._proxy_klasses->length(); i++) {
       InstanceKlass* ik = info._proxy_klasses->at(i);
-      if (!ik->can_be_verified_at_dumptime()) {
+      if (SystemDictionaryShared::check_for_exclusion_impl(ik, true)) {
```

Comment:
With the original code, if the list has two elements, only the first one is removed -- after the first iteration, `i` becomes 1, but `_proxy_klasses->length()` is also 1, and thus the loop will terminate.

I think the loop should be reversed:
```
for (int i = info._proxy_klasses->length() - 1; i >= 0; i--) {
  if (...) {
    info._proxy_klasses->remove_at(i);
  }
}
```

For test coverage, we should also have two test case:

- caller class Foo1 that uses a single instance of an lambda based on old interface
- caller class Foo2 that uses two instances of an lambda based on old interface

Reply:
Nice catch! I've fixed the code.
I've expanded the existing LambdaContainsOldInf.java test to cover the above cases.